src/cmd/flux-mini.py

```diff
@@ -0,0 +1,305 @@
+                raise ValueError("node count must be an integer >= 1 (if set)")
+            if num_nodes > num_tasks:
+                raise ValueError("node count must not be greater than task count")
+        core = self.__create_resource("core", cores_per_task)
+        if num_nodes is not None:
+            num_slots = int(math.ceil(num_tasks / float(num_nodes)))
+            if num_tasks % num_nodes != 0:
+                # N.B. uneven distribution results in wasted task slots
+                task_count_dict = {"total": num_tasks}
+            else:
+                task_count_dict = {"per_slot": 1}
+            slot = self.__create_slot("task", num_slots, [core])
+            resource_section = self.__create_resource("node", num_nodes, [slot])
+        else:
+            task_count_dict = {"per_slot": 1}
+            slot = self.__create_slot("task", num_tasks, [core])
+            resource_section = slot
+
+        self.jobspec = {
+            "version": 1,
+            "resources": [resource_section],
+            "tasks": [
+                {
+                    "command": command,
+                    "slot": "task",
+                    "count": task_count_dict,
+                    "attributes": {},
+                }
+            ],
+            "attributes": {"system": {"duration": 0}},
+        }
+
+    def __create_resource(self, res_type, count, with_child=[]):
+        assert isinstance(
+            with_child, collectionsAbc.Sequence
+        ), "child resource must be a sequence"
+        assert not isinstance(with_child, str), "child resource must not be a string"
+        assert count > 0, "resource count must be > 0"
+
+        res = {"type": res_type, "count": count}
+
+        if len(with_child) > 0:
+            res["with"] = with_child
+        return res
+
+    def __create_slot(self, label, count, with_child):
+        slot = self.__create_resource("slot", count, with_child)
+        slot["label"] = label
+        return slot
+
+    def __parse_fsd(self, s):
+        m = re.match(r".*([smhd])$", s)
+        try:
+            n = float(s[:-1] if m else s)
+        except:
+            raise ValueError("invalid Flux standard duration")
+        unit = m.group(1) if m else "s"
+
+        if unit == "m":
+            seconds = timedelta(minutes=n).total_seconds()
+        elif unit == "h":
+            seconds = timedelta(hours=n).total_seconds()
+        elif unit == "d":
+            seconds = timedelta(days=n).total_seconds()
+        else:
+            seconds = n
+        if seconds < 0 or math.isnan(seconds) or math.isinf(seconds):
+            raise ValueError("invalid Flux standard duration")
+        return seconds
+
+    def set_duration(self, duration):
+        """
+        Assign a time limit to the job.  The duration may be:
+        - a float in seconds
+        - a string in Flux Standard Duration
+        A duration of zero is interpreted as "not set".
+        """
+        if type(duration) is str:
```

Comment:
python2 and python3 compatible string handling is unfortunately a bit of a bear.  To make this compatible with python2, you'll need to check `if isinstance(duration, six.string_types)` [[1]](https://six.readthedocs.io/#six.string_types).

Reply:
Since python 2 EOL is coming imminently, it might be worth checking with our customers to see how long we need to hang on to python 2 support.  Alternately, we may require python3 and optionally additionally build python 2 bindings so we don’t have to deal with this stuff in our own commands.

Reply:
That's a great point @trws:  just opened an issue to track the python2 EOL actions.